drivers/platform/mbed/mbed_uart.cpp

```diff
@@ -0,0 +1,492 @@
+		usb_cdc_dev->change_terminal_connection(true);
+
+		usb_cdc_dev->receive_nb(data, bytes_number, &size_rd);
+		return bytes_number;
+	} else {
+		uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+						   desc->extra))->uart_port);
+		return uart->read(data, bytes_number);
+	}
+}
+
+/**
+ * @brief Write data to UART device.
+ * @param desc[in] - Instance of UART.
+ * @param data[in, out] - Pointer to buffer containing data.
+ * @param bytes_number[in] - Number of bytes to read.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t uart_write(struct uart_desc *desc, const uint8_t *data,
+		   uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;	// pointer to BufferedSerial/UART instance
+	platform_usbcdc *usb_cdc_dev;	// Pointer to usb cdc device class instance
+	uint32_t d_sz;
+	uint32_t indx = 0;
+	uint32_t tx_bytes = bytes_number;
+
+	if (!desc || !desc->extra || !data)
+		return -EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	if (((struct mbed_uart_desc *)desc->extra)->virtual_com_enable) {
+		usb_cdc_dev = (platform_usbcdc *)((struct mbed_uart_desc *)(
+				desc->extra))->uart_port;
+
+		while (bytes_number) {
+			while (!usb_cdc_dev->data_transmited()) {
+				/* Wait until old data is transmitted */
+			}
+
+			/* Make sure packet size is less than max CDC packet size during data transmit */
+			d_sz = (bytes_number > (USB_CDC_MAX_PACKET_SIZE - 1)) ?
+			       (USB_CDC_MAX_PACKET_SIZE - 1) :
+			       bytes_number;
+
+			/* Change terminal connection status manually */
+			usb_cdc_dev->change_terminal_connection(true);
+
+			usb_cdc_dev->send_nb((uint8_t *)&data[indx], d_sz, &d_sz);
+
+			bytes_number -= d_sz;
+			indx += d_sz;
+		}
+
+		return tx_bytes;
+	} else {
+		uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+						   desc->extra))->uart_port);
+		return uart->write(data, bytes_number);
+	}
+}
+
+/**
+ * @brief Submit reading buffer to the UART driver.
+ * Buffer is used until bytes_number bytes are read.
+ * @param desc[in]- Descriptor of the UART device
+ * @param data[out] - Buffer where data will be read
+ * @param bytes_number[in] - Number of bytes to be read.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Currently implemented only for UART and not for USBCDC (VCOM)
+ */
+int32_t uart_read_nonblocking(struct uart_desc *desc,
+			      uint8_t *data,
+			      uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;	// pointer to BufferedSerial/UART instance
+	uint32_t i;
+
+	if (!desc || !desc->extra || !data)
+		return -EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+					   desc->extra))->uart_port);
+
+	/* Return 0 bytes if nothing to read */
+	if (!uart->readable())
+		return 0;
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart->readable())
+			uart->read(&data[i], 1);
```

Comment:
I'd say we should `break` on the `else` path and return `i` instead of `bytes_number`. Because the danger here is that the user will ask for, let's say 5 bytes, and the buffer has 2 bytes. The function will read 2 bytes and then stay a little while more in the function and return that it has read 5 bytes, which might cause errors.

Reply:
Make sense. I'll update the code.